src/net_doh.c

```diff
@@ -0,0 +1,1140 @@
+
+    case NGHTTP2_RST_STREAM:
+        // debugx(": RST_STREAM\n");
+        break;
+    case NGHTTP2_GOAWAY:
+        // debugx(": GOAWAY\n");
+        break;
+    }
+    
+    return 0;
+}
+
+static int _http2_data_chunk_recv_cb(nghttp2_session* session, 
+                                     uint8_t flags, 
+                                     int32_t stream_id, 
+                                     const uint8_t* data, 
+                                     size_t len, void* user_data)
+{
+    perf__doh_socket_t *sock = (perf__doh_socket_t *)user_data;
+    (void)flags;
+
+    int ret;
+    // debugx("data_chunk_recv_cb");
+    if (nghttp2_session_get_stream_user_data(session, stream_id)) {
+        // debugx("data_chunk length: %d\n", len);
+        // fwrite(data, 1, len, stderr);
+        // debugx("\n");
+
+        if (self->http2->dnsmsg_at == 0) {
+            // TODO: upper boundary check
+            memcpy(self->http2->dnsmsg, data, len);
+            self->http2->dnsmsg_at = len;
+        } else {
+            // TODO: boundary checks for self->recvbuf_at + len
+            memcpy(self->http2->dnsmsg + self->http2->dnsmsg_at, data, len);
+            self->http2->dnsmsg_at += len;
+        }
+    }
+    
+    // TODO: read incoming data and append to buf
+    return 0;
+}
+
+static int _http2_init(struct perf__doh_socket* sock)
+{
+    debugx("_http2_init for qid: %d", sock->qid);
+    struct URI uri;
+    int ret = -1;
+    nghttp2_session_callbacks* callbacks;
+    nghttp2_option* option;
+
+    ret = parse_uri(&uri, net_doh_uri);
+
+    if (ret != 0) {
+        perf_log_fatal("Failed to parse https URI");
+    }
+  
+    self->http2 = http2_session_init();
+    self->http2->stream = http2_stream_init(&uri);
+    self->http2->max_concurrent_streams = DEFAULT_MAX_CONCURRENT_STREAMS;
+
+    /* sets HTTP/2 callbacks */
+    assert(nghttp2_session_callbacks_new(&callbacks) == 0);
+    nghttp2_session_callbacks_set_send_callback(callbacks, _http2_send_cb);
+    // nghttp2_session_callbacks_set_recv_callback(callbacks, _http2_recv_cb); // TODO: remove past debug
+    nghttp2_session_callbacks_set_on_header_callback(callbacks, _http2_header_cb);
+    nghttp2_session_callbacks_set_on_data_chunk_recv_callback(callbacks, _http2_data_chunk_recv_cb);
+    nghttp2_session_callbacks_set_on_frame_recv_callback(callbacks, _http2_frame_recv_cb);
+    nghttp2_session_callbacks_set_on_frame_send_callback(callbacks, _http2_frame_send_cb); // TODO: remove - debug
+    nghttp2_session_callbacks_set_on_stream_close_callback(callbacks, _http2_stream_close_cb);
+
+    assert(nghttp2_option_new(&option) == 0);
+    nghttp2_option_set_peer_max_concurrent_streams(option, self->http2->max_concurrent_streams);
+
+    ret = nghttp2_session_client_new2(&self->http2->session, callbacks, self, option);
+
+    nghttp2_session_callbacks_del(callbacks);
+    nghttp2_option_del(option);
+
+    debugx("http2 session client ret: %d", ret);
+    if (ret < 0) {
```

Comment:
need better handling of return by `nghttp2_session_client_new2()`, following code will segfault!

Reply:
thanks for this, yes - _client_new2() will do 0 or -901 ERR_NOMEM which is fatal, will amend.